uavcan/register/384.Access.1.0.uavcan

```diff
@@ -64,19 +64,68 @@
 # The values of registers containing non-scalar numerical entities should be compared elementwise.
 #
 # The following table specifies the register name patterns that are reserved by the specification for
-# common functions. Implementers should always follow these conventions whenever applicable.
-# The table uses the following abbreviations for register flags:
-#   - M - mutable, I - immutable
-#   - P - persistent, V - volatile
-#
-#   Name pattern        | Flags | Type      | Purpose
-# ----------------------+-------+-----------+--------------------------------------------------------------------------
-#   uavcan.node_id      | MP    | uint16    | Contains the node-ID of the local node. Values above the maximum valid
-#                       |       |           | node-ID for the current transport indicate that the node-ID is not set;
-#                       |       |           | if plug-and-play is supported, it will be used by the node to obtain an
-#                       |       |           | automatic node-ID. Invalid values other than 65535 should be avoided for
-#                       |       |           | consistency. The factory-default value should be 65535.
-# ----------------------+-------+-----------+--------------------------------------------------------------------------
+# common functions. These conventions are not mandatory to follow, but implementers are recommended to adhere because
+# they enable enhanced introspection capabilities and simplify device configuration and diagnostics.
+#
+#   REGISTER NAME PATTERN                             TYPE            FLAGS                   RECOMMENDED DEFAULT
+# =====================================================================================================================
+#
+#   uavcan.node.id                                      natural16       mutable, persistent     65535 (unset/PnP)
+#
+# Contains the node-ID of the local node. Values above the maximum valid node-ID for the current transport
+# indicate that the node-ID is not set; if plug-and-play is supported, it will be used by the node to obtain an
+# automatic node-ID. Invalid values other than 65535 should be avoided for consistency.
+#
+# ---------------------------------------------------------------------------------------------------------------------
+#
+#   uavcan.node.description                             string          mutable, persistent     (empty)
+#
+# Human-readable verbose description of this specific node defined by the integrator (not vendor).
+#
+# ---------------------------------------------------------------------------------------------------------------------
+#
+#   uavcan.cookie                                       string          mutable, persistent     (empty)
```

Comment:
as the name suggests, this concept risks becoming a malware enabler depending on how systems decide to use, or abuse, this storage facility. Why do we need this?

Reply:
> this concept risks becoming a malware enabler

In the same way as a kitchen knife is a cold weapon, yes. The very same thing can be said about the bootloader protocol, the LAN forwarding interface, and just about every other feature of UAVCAN.

There is currently no cybersecurity measures, and I am not exactly clear on how this side of the project will be evolving in the future. Conventional aerospace systems ensure (cyber)security by preventing unauthorized access to the aircraft's systems; this model may no longer be applicable in the future due to certain new capabilities of modern systems but it's a different story. In my humble opinion, the overall perception of the importance of cybersec in aerospace is somewhat artificially inflated mostly because people experienced in ICT mistakenly carry over core assumptions into a different field where they are no longer valid. This occasionally leads to [appalingly useless security research publications](https://www.rapid7.com/research/report/investigating-can-bus-network-integrity-in-avionics-systems/) that can be effectively summarized as "our new research shows that removing the wing breaks the airplane". At the moment, I assign a low priority to this matter, but it doesn't mean that it is out of the scope entirely or that it is not something that will be worked on. There are far more critical issues out there.

![image](https://user-images.githubusercontent.com/3298404/97357329-41a47e80-18a2-11eb-9d63-c8b536c793af.png)

Currently, cybersec is backlogged in DS-015 with a note that it may be addressed either in DS-015 itself or in the core UAVCAN Specification, depending on how things develop.

Speaking about this cookie register specifically: it will be relied on by the upcoming autoconfiguration algorithm; it is based on this post in an early revision of the DS-015 document:

> "PX4 node gets the heartbeat..." this requires storing persistent data in a 'node id table'. Storage means we have to worry about maximum sizes of tables (i.e. how much non-volatile memory must your hardware provide to host this function) and consistency (i.e. how do you know the data hasn't become corrupt).
> 
> Another strategy is to require each node to remember its own "I am configured" state. This provides a bounded storage problem distributed to each node. It's also easier for the node to monitor consistency and if there is a failure of its persistent storage you don't lose the entire database; just that one node becomes unconfigured. Perhaps this activity can be redesigned to use some sort of "system ID" where the function can be distributed (Pavel loves distributed!). In such a scheme there is a distribution on the bus of "this is the system ID you are attached to" and if a Node sees it hasn't configured for that system it can enter a system configuration activity. This activity might be the node looking at it's registers and for each that doesn't have a port identifier received from this system it broadcasts "I need to know about port ID X for system Y". The code cannot become operational until all port identifiers it requires for its primary function have been provided to it. It would seem this allows attached devices to better discover and adapt to system capabilities. For example, if I ask for the port identifier for function "myfunction_2_0" and nobody knows I can then ask "does anybody have port identifiers for myfunction_1_0?," etc.
> 
> In this scheme the prompt to a user for "there's an unknown node" would be a function that simply monitored for any node that did not report it was configured for the current system.
> 
> So, this is just some quick thoughts here. Sorry I don't have time to flush them out but in-general I think you need to find a way to remove the central "configured nodes" database here that could be a single point of failure for the system.

This change request is a bit sudden and it is not backed up by a comprehensive proposal. I am cutting corners to ensure that we will be able to retrofit autoconfiguration in DS-015 v1.x (x > 0) without the need to ever touch peripheral node implementations that conform to DS-015 v1.0. If there are objections against such (low-risk) deviation from our normal design process, we can pull this specific register from the changeset, but keep in mind that requiring implementers of DS-015 v1.0 to change the firmware to support autoconfiguration may do damage to the ecosystem.

Reply:
my comment from the forum was suggesting a read-only register for "I am configured" though. My question is, why is this an externally _writable_ register and not an immutable, volatile register?

Reply:
The register is expected to store a device-specific context that is entirely determined by the *autoconfiguration authority* that writes and interprets its content. The objective is to relieve said authority from mapping the node identity with its specific settings internally. If you want to discuss this at length, I think it is best to pull this change from this PR and accept the rest, postponing this issue until the point where the DS-015 autoconfiguration proposal is ready.

Reply:
I'm not vehemently opposed to this but I'm not able to really review it since I don't understand the context. It might be best to pull it and reintroduce when that context is available.